File: simplex.java/src/main/java/kr/co/tmon/simplex/actions/AbstractAction.java
package kr.co.tmon.simplex.actions;

import kr.co.tmon.simplex.channels.Channel;
import kr.co.tmon.simplex.channels.ChannelZip;
import kr.co.tmon.simplex.channels.IChannel;

public abstract class AbstractAction<TParam, TResult> implements IAction<TParam, TResult> {

	private IChannel defaultChannel;
	
	@Override
	public IChannel getDefaultChannel() {
		return defaultChannel;
	}
	
	@Override
	public TransformedResult<TResult> transform(Throwable throwable) {
		return new TransformedResult<TResult>(false, null);
	}

	@Override
	public IChannel zip(IChannel...channels) {
		IChannel channel = new ChannelZip(channels);
		return channel;
	}
	
	public AbstractAction() {
		defaultChannel = new Channel();
	}
}

File: Simplex.kotlin/src/main/kotlin/kr/co/tmon/simplex/actions/AbstractAction.kt
package kr.co.tmon.simplex.actions

import kr.co.tmon.simplex.channels.*
import kr.co.tmon.simplex.extensions.*
import io.reactivex.Observable

abstract class AbstractAction<TParam, TResult> : IAction<TParam, TResult> {

	internal val default: IChannel by lazy { Channel() }

	override fun transform(exception: Throwable) : TransformedResult<TResult> {
		return TransformedResult(false, null)
	}

	abstract override fun process(param: TParam?) : Observable<TResult>

	override fun zip(vararg channels: IChannel) : IChannel {
		return channels.zip()
	}
}

